### PR TITLE
[WIP] RCHAIN-3182

### PR DESCRIPTION
**Pull request by JosDenmark**
### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3182

### CODE REVIEW

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    extends RhoSpec(
+      CompiledRholangSource("PoSTest.rho"),
+      Seq(
+        StandardDeploys.nonNegativeNumber,
+        StandardDeploys.makeMint,
+        StandardDeploys.authKey,
+        StandardDeploys.either,
+        StandardDeploys.revVault,
+        StandardDeploys.PoS
+      ),
```
Comment (KentShikama) [inline casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala]: You'll probably need to change this based on https://github.com/rchain/rchain/pull/2328

Comment (JosDenmark) [reply casper/src/test/scala/coop/rchain/casper/genesis/contracts/PoSSpec.scala]: Done.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala [hunk 2]
```diff
@@ -136,7 +137,7 @@ object SystemProcesses {
           val response =
             RevAddress
               .fromPublicKey(PublicKey(publicKey))
-              .map(RhoType.String(_))
+              .map(ra => RhoType.String(Base58.encode(ra.keyHash)))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: Can we have a `.toBase58` method on `Address`?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala [hunk 3]
```diff
@@ -16,13 +16,13 @@ class AddressTools(prefix: Array[Byte], keyLength: Int, checksumLength: Int) {
...
     */
-  def fromPublicKey(pk: PublicKey): Option[String] =
+  def fromPublicKey(pk: PublicKey): Option[Address] =
     if (keyLength == pk.bytes.length) {
       val keyHash = Blake2b256.hash(pk.bytes)
       val payload = prefix ++ keyHash
       val address = payload ++ computeChecksum(payload)
 
-      Some(Base58.encode(address))
+      Some(Address(prefix, address))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala]: you're putting the whole address (prefix + keyHash + checksum) on the Address.keyHash field here.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala]: We should return `Some(Adddress(prefix, keyHash)` and have

```scala
//Address class
def toBase58() = {
    val payload = prefix ++ keyHash
    val address = payload ++ computeChecksum(payload)	      val address = payload ++ computeChecksum(payload)
	address
}
```

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/Validator.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package coop.rchain.casper.genesis.contracts
+
+import coop.rchain.crypto.PublicKey
+import coop.rchain.models.Par
+import coop.rchain.rholang.build.CompiledRholangSource
+import coop.rchain.rholang.interpreter.ParBuilder
+import monix.eval.Coeval
+
+// TODO: Eliminate public key argument if unnecessary
+final case class Validator(pk: PublicKey, stake: Long) extends CompiledRholangSource {
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/genesis/contracts/Validator.scala]: This commit says `Extract Validator data type into separate file.`

Yet, it introduces a new `Validator.scala` and removes `VaultUtils.scala`. Bad merge?

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/RevGenerator.scala [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import coop.rchain.rholang.build.CompiledRholangSource
+import coop.rchain.rholang.interpreter.ParBuilder
+import coop.rchain.rholang.interpreter.util.RevAddress
+import coop.rchain.rholang.interpreter.util.codec.Base58
+import monix.eval.Coeval
+
+final case class RevGenerator(genesisAddress: RevAddress, userVaults: Seq[Vault], supply: Long)
+    extends CompiledRholangSource {
+
+  val path: String = "<synthetic in Rev.scala>"
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/genesis/contracts/RevGenerator.scala]: `<synthetic in RevGenerator.scala>`

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/StandardDeploys.scala [hunk 6]
```diff
@@ -107,4 +110,12 @@ object StandardDeploys {
       "be8ae000e4d2f29c73f792705314d71b6c0d56d7c640c6b4df9fabf90518c623",
       1551879405043L
     )
+
+  def testRev(genesisPk: PublicKey, vaults: Seq[Vault], supply: Long): DeployData =
+    toDeploy(
+      RevGenerator(RevAddress.fromPublicKey(genesisPk).get, vaults, supply),
+      s"${Base16.encode(genesisPk.bytes)}",
```
Comment (ArturGajowy) [inline casper/src/main/scala/coop/rchain/casper/genesis/contracts/StandardDeploys.scala]: a `.toBase16` method on PrivateKey and PublicKey would help me a lot when reading such code, and probably would be handy when debugging (easier logging / setting up a watch)

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/TestUtil.scala [hunk 7]
```diff
@@ -93,4 +101,82 @@ object TestUtil {
...
+    val defaultDeploys =
+      defaultTerms(
+        genesisPk,
+        vaults,
+        revSupply,
+        minimumBond,
+        maximumBond
+      ) ++ bondDeploys
+
+    Genesis.withContracts(
```
Comment (ArturGajowy) [inline casper/src/test/scala/coop/rchain/casper/genesis/contracts/TestUtil.scala]: The growth of this test setup worries me. I'd rather imagine the deploys to be computed inside the Genesis methods, based on arguments like `genesisPubKey` or `vaults`. One consequence of having this duplicated is - to state the obvious - there's nothing that keeps it in sync with how it behaves in prod.

File: casper/src/main/resources/PoS.rho [hunk 8]
```diff
@@ -16,9 +16,9 @@
 
 new PoS, rs(`rho:registry:insertSigned:ed25519`), uriOut, stdout(`rho:io:stdout`) in {
 
-  for (@minimumBond, @maximumBond, @initialBonds <- PoS) {
+ for (@minimumBond, @maximumBond <- PoS) {
 
-    match "11112GofKy1v5mPWZjZEYBByvpZ4ahYJp6vFN1hgGCn7L543rFSpyA".hexToBytes() {
+ match "11112GofKy1v5mPWZjZEYBByvpZ4ahYJp6vFN1hgGCn7L543rFSpyA" {
```
Comment (ArturGajowy) [inline casper/src/main/resources/PoS.rho]: I think this change should be part of the original commit introducing this line, or better described

File: casper/src/main/resources/PoS.rho [hunk 9]
```diff
@@ -16,9 +16,9 @@
 
 new PoS, rs(`rho:registry:insertSigned:ed25519`), uriOut, stdout(`rho:io:stdout`) in {
 
-  for (@minimumBond, @maximumBond, @initialBonds <- PoS) {
+ for (@minimumBond, @maximumBond <- PoS) {
```
Comment (ArturGajowy) [inline casper/src/main/resources/PoS.rho]: This should be in the commit where we stop passing the initial bonds

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala (hunk 2 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/SystemProcesses.scala]: Done in dev already

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala (hunk 3 above)
Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/util/AddressTools.scala]: Done in dev

**Pull request closed without merging**